metals/src/main/scala/scala/meta/internal/metals/codeactions/CreateCompanionObjectCodeAction.scala

```diff
@@ -0,0 +1,134 @@
+package scala.meta.internal.metals.codeactions
+
+import org.eclipse.lsp4j.{CodeActionParams, Location}
+import org.eclipse.{lsp4j => l}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+import scala.meta.Defn
+
+import scala.meta.internal.metals.ServerCommands
+
+/**
+ * look at the
+ *
+ * @param trees
+ */
+class CreateCompanionObjectCodeAction(
+    trees: Trees
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    pprint.log("contribute")
+    val uri = params.getTextDocument().getUri()
+    val path = uri.toAbsolutePath
+    val range = params.getRange()
+    val applyTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Defn.Class](
+            path,
+            range.getStart(),
+            applyWithSingleFunction
+          )
+      else None
+
+    pprint.log(applyTree)
+
+    applyTree.collect { case classDefinition: Defn.Class =>
+      findCompanionObjectOfClass(classDefinition) match {
+        case Some(comanionObject) =>
+          buildShowingCompanionObjectCodeAction(comanionObject, uri)
+        case None =>
+          buildCreatingCompanionObjectCodeAction(path, classDefinition, uri)
+      }
+    }.toSeq
+
+  }
+
+  def buildCreatingCompanionObjectCodeAction(
+      path: AbsolutePath,
+      classDefinition: Defn.Class,
+      uri: String
+  ): l.CodeAction = {
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(CreateCompanionObjectCodeAction.companionObjectInfo)
+    codeAction.setKind(this.kind)
+    val rangeStart = classDefinition.pos.toLSP.getEnd
+    val rangeEnd = classDefinition.pos.toLSP.getEnd
+    rangeEnd.setLine(rangeEnd.getLine + 5)
```

Comment:
This might break the code if there are other classes afterwards. Instead of adding the line we can prepend `companionObjectFirstLine` with newlines

Reply:
Sorry, I am not sure if I understand this! How do I specify the range end then? How do I prepend new lines?

Reply:
Great! Let me try!